File: backend-go/internal/domain/agregado_assembly.go
package domain

import (
	"errors"
	"fmt"
	"strings"
)

// RigidMemberSourceKind discriminates whether a rigid member points directly to a fixed
// hardware ID or dynamically selects from a product variant set.
type RigidMemberSourceKind string

const (
	RigidMemberSourceFixed   RigidMemberSourceKind = "fixed"
	RigidMemberSourceVariant RigidMemberSourceKind = "variant"
)

// RigidMemberBOMRole defines how a rigid member participates in cost and purchasing.
type RigidMemberBOMRole string

const (
	// BOMRoleIncludedInKit: Member is physically included inside the commercial kit box.
	// It does NOT produce an individual purchasing line item in the BOM.
	BOMRoleIncludedInKit RigidMemberBOMRole = "included_in_kit"
	// BOMRoleSeparatelyPurchased: Member is an additional component purchased individually.
	// It produces its own purchasing line item in the BOM.
	BOMRoleSeparatelyPurchased RigidMemberBOMRole = "separately_purchased"
	// BOMRoleNonPurchasing: Member is a visual representation only (e.g. external reference).
	// It produces no purchasing line item.
	BOMRoleNonPurchasing RigidMemberBOMRole = "non_purchasing"
)

// AssemblyAxisRef defines the reference point along an axis of the assembly bounding box.
type AssemblyAxisRef string

const (
	AxisRefMin    AssemblyAxisRef = "min"    // minX = 0, minY = 0, minZ = 0
	AxisRefMax    AssemblyAxisRef = "max"    // maxX = W, maxY = D, maxZ = H
	AxisRefCenter AssemblyAxisRef = "center" // centerX = W/2, centerY = D/2, centerZ = H/2
)

// AssemblyAxisPlacement positions a coordinate along an axis via a reference point and offset in mm.
type AssemblyAxisPlacement struct {
	Ref      AssemblyAxisRef `json:"ref"`      // "min" | "max" | "center"
	OffsetMm float64         `json:"offsetMm"` // signed delta in mm from the reference
}

// AssemblyAnchorRule defines the rigid pose of a member inside the Assembly Space.
type AssemblyAnchorRule struct {
	X           AssemblyAxisPlacement `json:"x"`
	Y           AssemblyAxisPlacement `json:"y"`
	Z           AssemblyAxisPlacement `json:"z"`
	RotationDeg *HardwareRotationDeg  `json:"rotationDeg,omitempty"` // Rigid rotation (multiples of 90 deg: 0, 90, 180, 270)
}

// FixedHardwareSource points to a single fixed Hardware definition.
type FixedHardwareSource struct {
	HardwareID string `json:"hardwareId"`
}

// VariantHardwareSource points to an assembly-scoped VariantSet.
type VariantHardwareSource struct {
	VariantSetID string `json:"variantSetId"`
}

// RigidMemberSource is the discriminated union for hardware binding. Exactly one source must be populated.
type RigidMemberSource struct {
	Kind    RigidMemberSourceKind  `json:"kind"`
	Fixed   *FixedHardwareSource   `json:"fixed,omitempty"`
	Variant *VariantHardwareSource `json:"variant,omitempty"`
}

// AgregadoRigidMember defines an autonomous bought rigid hardware component inside an Agregado recipe.
type AgregadoRigidMember struct {
	MemberID  string             `json:"memberId"`
	Role      string             `json:"role"` // e.g. "side_left", "side_right", "runner_left", "runner_right"
	Source    RigidMemberSource  `json:"source"`
	Placement AssemblyAnchorRule `json:"placement"`
	BOMRole   RigidMemberBOMRole `json:"bomRole"`
}

// ProductVariant represents one discrete commercial option in a VariantSet.
type ProductVariant struct {
	NominalDimensionMm float64 `json:"nominalDimensionMm"` // e.g. 450.0, 500.0, 550.0
	HardwareID         string  `json:"hardwareId"`         // Catalog Hardware.ID with exact 3D Asset & MountFrame
}

// AgregadoVariantSet groups discrete commercial sizes available for a specific dimension.
type AgregadoVariantSet struct {
	ID        string           `json:"id"`
	Dimension string           `json:"dimension"` // "depth" | "height" | "width"
	Variants  []ProductVariant `json:"variants"`
}

// AssemblyCompatibilityRule specifies how the Agregado recipe selects and validates variants.
type AssemblyCompatibilityRule struct {
	VariantSetID      string  `json:"variantSetId"`
	ClearanceMm       float64 `json:"clearanceMm"`       // Minimum space clearance required beyond nominal dimension
	SelectionStrategy string  `json:"selectionStrategy"` // "max_fitting" | "exact"
}

// AssemblyMemberTransform defines the authoritative translation and orthonormal basis of a member.
type AssemblyMemberTransform struct {
	TranslationMm [3]float64    `json:"translationMm"`
	Basis         HardwareBasis `json:"basis"`
}

// ResolvedRigidMember is the fully resolved rigid hardware member ready for 3D placement.
type ResolvedRigidMember struct {
	MemberID        string                  `json:"memberId"`
	Role            string                  `json:"role"`
	HardwareID      string                  `json:"hardwareId"`
	AssetID         string                  `json:"assetId"`
	AssetRevisionID string                  `json:"assetRevisionId"`
	SHA256          string                  `json:"sha256"`
	MountFrame      *HardwareMountFrame     `json:"mountFrame,omitempty"`
	LocalTransform  AssemblyMemberTransform `json:"localTransform"` // Rigid placement in Assembly Space (det = +1.0)
	BOMRole         RigidMemberBOMRole      `json:"bomRole"`
}

// AssemblyBOMItem is one purchasing row produced by the assembly.
type AssemblyBOMItem struct {
	HardwareID string  `json:"hardwareId"`
	Quantity   float64 `json:"quantity"`
	Role       string  `json:"role"`
	Notes      string  `json:"notes,omitempty"`
}

// ResolvedFabricatedComponent describes a dimensioned and placed fabricated board.
type ResolvedFabricatedComponent struct {
	ComponentID string                  `json:"componentId"`
	SlotID      string                  `json:"slotId"`
	Name        string                  `json:"name"`
	LengthMm    float64                 `json:"lengthMm"`
	WidthMm     float64                 `json:"widthMm"`
	ThicknessMm float64                 `json:"thicknessMm"`
	Transform   AssemblyMemberTransform `json:"transform"`
}

// ResolvedAssemblySnapshot freezes all resolved state for immutable historical DesignRevisions.
type ResolvedAssemblySnapshot struct {
	AgregadoID              string                        `json:"agregadoId"`
	AgregadoRevisionNumber  int                           `json:"agregadoRevisionNumber"`
	CommercialKitHardwareID *string                       `json:"commercialKitHardwareId,omitempty"`
	ResolvedDimensionsMm    [3]float64                    `json:"resolvedDimensionsMm"`
	SelectedVariants        map[string]float64            `json:"selectedVariants"` // variantSetId -> nominalDimensionMm
	RigidMembers            []ResolvedRigidMember         `json:"rigidMembers"`
	FabricatedComponents    []ResolvedFabricatedComponent `json:"fabricatedComponents"`
	BOMItems                []AssemblyBOMItem             `json:"bomItems"`
}

// Domain Errors
type ErrAssemblyVariantNotFound struct {
	VariantSetID       string
	RequestedSpaceMm   float64
	RequiredClearance float64
	AvailableNominals  []float64
}

func (e *ErrAssemblyVariantNotFound) Error() string {
	return fmt.Sprintf(
		"assembly variant not found for variantSetId '%s': requested space %.1fmm (clearance %.1fmm) cannot accommodate available nominals %v",
		e.VariantSetID, e.RequestedSpaceMm, e.RequiredClearance, e.AvailableNominals,
	)
}

// Validation Functions

// ValidateRigidMemberSource enforces strict discriminated union exclusivity.
func ValidateRigidMemberSource(s RigidMemberSource) error {
	switch s.Kind {
	case RigidMemberSourceFixed:
		if s.Fixed == nil || strings.TrimSpace(s.Fixed.HardwareID) == "" {
			return errors.New("fixed source requires non-empty hardwareId")
		}
		if s.Variant != nil {
			return errors.New("fixed source must not contain variant payload")
		}
	case RigidMemberSourceVariant:
		if s.Variant == nil || strings.TrimSpace(s.Variant.VariantSetID) == "" {
			return errors.New("variant source requires non-empty variantSetId")
		}
		if s.Fixed != nil {
			return errors.New("variant source must not contain fixed payload")
		}
	default:
		return fmt.Errorf("invalid rigid member source kind: '%s'", s.Kind)
	}
	return nil
}

// ValidateAxisPlacement verifies axis reference value.
func ValidateAxisPlacement(axisName string, p AssemblyAxisPlacement) error {
	switch p.Ref {
	case AxisRefMin, AxisRefMax, AxisRefCenter:
		return nil
	default:
		return fmt.Errorf("axis %s has invalid reference '%s' (expected min, max, or center)", axisName, p.Ref)
	}
}

// ValidateAssemblyAnchorRule validates placement references and rigid rotation.
func ValidateAssemblyAnchorRule(r AssemblyAnchorRule) error {
	if err := ValidateAxisPlacement("x", r.X); err != nil {
		return err
	}
	if err := ValidateAxisPlacement("y", r.Y); err != nil {
		return err
	}
	if err := ValidateAxisPlacement("z", r.Z); err != nil {
		return err
	}
	return nil
}

// ValidateAgregadoRigidMember validates an individual rigid member definition against kit policy.
func ValidateAgregadoRigidMember(m AgregadoRigidMember, hasCommercialKit bool) error {
	if strings.TrimSpace(m.MemberID) == "" {
		return errors.New("rigid member must have a non-empty memberId")
	}
	if strings.TrimSpace(m.Role) == "" {
		return fmt.Errorf("rigid member %s must have a non-empty role", m.MemberID)
	}
	if err := ValidateRigidMemberSource(m.Source); err != nil {
		return fmt.Errorf("rigid member %s: %w", m.MemberID, err)
	}
	if err := ValidateAssemblyAnchorRule(m.Placement); err != nil {
		return fmt.Errorf("rigid member %s: %w", m.MemberID, err)
	}

	switch m.BOMRole {
	case BOMRoleIncludedInKit:
		if !hasCommercialKit {
			return fmt.Errorf("rigid member %s: bomRole 'included_in_kit' is invalid when no commercialKitHardwareId is configured", m.MemberID)
		}
	case BOMRoleSeparatelyPurchased, BOMRoleNonPurchasing:
		// Always valid
	default:
		return fmt.Errorf("rigid member %s: invalid bomRole '%s'", m.MemberID, m.BOMRole)
	}

	return nil
}

// ValidateAgregadoVariantSet validates variant set consistency.
func ValidateAgregadoVariantSet(vs AgregadoVariantSet) error {
	if strings.TrimSpace(vs.ID) == "" {
		return errors.New("variant set must have a non-empty id")
	}
	switch vs.Dimension {
	case "depth", "height", "width":
		// valid
	default:
		return fmt.Errorf("variant set %s has invalid dimension '%s' (expected depth, height, or width)", vs.ID, vs.Dimension)
	}
	if len(vs.Variants) == 0 {
		return fmt.Errorf("variant set %s must define at least one ProductVariant", vs.ID)
	}
	for i, v := range vs.Variants {
		if v.NominalDimensionMm <= 0 {
			return fmt.Errorf("variant set %s variant %d: nominalDimensionMm must be positive", vs.ID, i)
		}
		if strings.TrimSpace(v.HardwareID) == "" {
			return fmt.Errorf("variant set %s variant %d: hardwareId must be non-empty", vs.ID, i)
		}
	}
	return nil
}

File: packages/domain/src/agregadoAssembly.ts
/**
 * Agregado Hardware Assembly Domain Contract
 *
 * Extends Agregado sub-assemblies with multi-member parametric hardware systems
 * (e.g. drawer runner systems like Blum MERIVOBOX, lift systems, sliding systems).
 *
 * Invariants:
 * 1. ZERO PARAMETRIC SCALING for rigid members: RigidMember geometry scale = [1,1,1] with det = +1.0.
 * 2. Fabricated members (bottom, back, etc.) recalculate dimensions from formulas and regenerate geometry.
 * 3. Assembly space follows workshop frame (+X width/right, +Y depth/front, +Z height/up).
 * 4. Distinct left/right parts are modeled as separate rigid members (never mirrored with negative scale).
 * 5. Strict discriminated union for rigid member source (fixed vs variant).
 */

export interface HardwareRotationDeg {
  readonly x?: number;
  readonly y?: number;
  readonly z?: number;
}

export type RigidMemberSourceKind = 'fixed' | 'variant';

export type RigidMemberBOMRole =
  | 'included_in_kit'
  | 'separately_purchased'
  | 'non_purchasing';

export type AssemblyAxisRef = 'min' | 'max' | 'center';

export interface AssemblyAxisPlacement {
  readonly ref: AssemblyAxisRef;
  readonly offsetMm: number;
}

export interface AssemblyAnchorRule {
  readonly x: AssemblyAxisPlacement;
  readonly y: AssemblyAxisPlacement;
  readonly z: AssemblyAxisPlacement;
  readonly rotationDeg?: HardwareRotationDeg;
}

export interface FixedHardwareSource {
  readonly hardwareId: string;
}

export interface VariantHardwareSource {
  readonly variantSetId: string;
}

export interface RigidMemberSource {
  readonly kind: RigidMemberSourceKind;
  readonly fixed?: FixedHardwareSource;
  readonly variant?: VariantHardwareSource;
}

export interface AgregadoRigidMember {
  readonly memberId: string;
  readonly role: string;
  readonly source: RigidMemberSource;
  readonly placement: AssemblyAnchorRule;
  readonly bomRole: RigidMemberBOMRole;
}

export interface ProductVariant {
  readonly nominalDimensionMm: number;
  readonly hardwareId: string;
}

export interface AgregadoVariantSet {
  readonly id: string;
  readonly dimension: 'depth' | 'height' | 'width';
  readonly variants: readonly ProductVariant[];
}

export interface AssemblyCompatibilityRule {
  readonly variantSetId: string;
  readonly clearanceMm: number;
  readonly selectionStrategy: 'max_fitting' | 'exact';
}

export interface AssemblyBasis {
  readonly x: readonly [number, number, number];
  readonly y: readonly [number, number, number];
  readonly z: readonly [number, number, number];
}

export interface AssemblyMemberTransform {
  readonly translationMm: readonly [number, number, number];
  readonly basis: AssemblyBasis;
}

export interface ResolvedRigidMember {
  readonly memberId: string;
  readonly role: string;
  readonly hardwareId: string;
  readonly assetId: string;
  readonly assetRevisionId: string;
  readonly sha256: string;
  readonly mountFrame?: {
    readonly originMm: readonly [number, number, number];
    readonly basis: AssemblyBasis;
  };
  readonly localTransform: AssemblyMemberTransform;
  readonly bomRole: RigidMemberBOMRole;
}

export interface AssemblyBOMItem {
  readonly hardwareId: string;
  readonly quantity: number;
  readonly role: string;
  readonly notes?: string;
}

export interface ResolvedFabricatedComponent {
  readonly componentId: string;
  readonly slotId: string;
  readonly name: string;
  readonly lengthMm: number;
  readonly widthMm: number;
  readonly thicknessMm: number;
  readonly transform: AssemblyMemberTransform;
}

export interface ResolvedAssemblySnapshot {
  readonly agregadoId: string;
  readonly agregadoRevisionNumber: number;
  readonly commercialKitHardwareId?: string;
  readonly resolvedDimensionsMm: readonly [number, number, number];
  readonly selectedVariants: Record<string, number>;
  readonly rigidMembers: readonly ResolvedRigidMember[];
  readonly fabricatedComponents: readonly ResolvedFabricatedComponent[];
  readonly bomItems: readonly AssemblyBOMItem[];
}

export interface AssemblyResolutionParams {
  readonly widthMm: number;
  readonly depthMm: number;
  readonly heightMm: number;
}

export class AssemblyVariantNotFoundError extends Error {
  readonly variantSetId: string;
  readonly requestedSpaceMm: number;
  readonly requiredClearance: number;
  readonly availableNominals: readonly number[];

  constructor(
    variantSetId: string,
    requestedSpaceMm: number,
    requiredClearance: number,
    availableNominals: readonly number[],
  ) {
    super(
      `assembly variant not found for variantSetId '${variantSetId}': requested space ${requestedSpaceMm.toFixed(1)}mm (clearance ${requiredClearance.toFixed(1)}mm) cannot accommodate available nominals [${availableNominals.join(', ')}]`,
    );
    this.name = 'AssemblyVariantNotFoundError';
    this.variantSetId = variantSetId;
    this.requestedSpaceMm = requestedSpaceMm;
    this.requiredClearance = requiredClearance;
    this.availableNominals = availableNominals;
  }
}

// Domain Validators

export function validateRigidMemberSource(source: RigidMemberSource): void {
  switch (source.kind) {
    case 'fixed':
      if (!source.fixed || !source.fixed.hardwareId.trim()) {
        throw new Error('fixed source requires non-empty hardwareId');
      }
      if (source.variant) {
        throw new Error('fixed source must not contain variant payload');
      }
      break;
    case 'variant':
      if (!source.variant || !source.variant.variantSetId.trim()) {
        throw new Error('variant source requires non-empty variantSetId');
      }
      if (source.fixed) {
        throw new Error('variant source must not contain fixed payload');
      }
      break;
    default:
      throw new Error(`invalid rigid member source kind: '${(source as { kind: string }).kind}'`);
  }
}

export function validateAxisPlacement(axisName: string, p: AssemblyAxisPlacement): void {
  if (p.ref !== 'min' && p.ref !== 'max' && p.ref !== 'center') {
    throw new Error(
      `axis ${axisName} has invalid reference '${p.ref}' (expected min, max, or center)`,
    );
  }
}

export function validateAssemblyAnchorRule(rule: AssemblyAnchorRule): void {
  validateAxisPlacement('x', rule.x);
  validateAxisPlacement('y', rule.y);
  validateAxisPlacement('z', rule.z);
}

export function validateAgregadoRigidMember(
  m: AgregadoRigidMember,
  hasCommercialKit: boolean,
): void {
  if (!m.memberId.trim()) {
    throw new Error('rigid member must have a non-empty memberId');
  }
  if (!m.role.trim()) {
    throw new Error(`rigid member ${m.memberId} must have a non-empty role`);
  }
  validateRigidMemberSource(m.source);
  validateAssemblyAnchorRule(m.placement);

  switch (m.bomRole) {
    case 'included_in_kit':
      if (!hasCommercialKit) {
        throw new Error(
          `rigid member ${m.memberId}: bomRole 'included_in_kit' is invalid when no commercialKitHardwareId is configured`,
        );
      }
      break;
    case 'separately_purchased':
    case 'non_purchasing':
      break;
    default:
      throw new Error(`rigid member ${m.memberId}: invalid bomRole '${m.bomRole}'`);
  }
}

export function validateAgregadoVariantSet(vs: AgregadoVariantSet): void {
  if (!vs.id.trim()) {
    throw new Error('variant set must have a non-empty id');
  }
  if (vs.dimension !== 'depth' && vs.dimension !== 'height' && vs.dimension !== 'width') {
    throw new Error(
      `variant set ${vs.id} has invalid dimension '${vs.dimension}' (expected depth, height, or width)`,
    );
  }
  if (!vs.variants || vs.variants.length === 0) {
    throw new Error(`variant set ${vs.id} must define at least one ProductVariant`);
  }
  for (let i = 0; i < vs.variants.length; i++) {
    const v = vs.variants[i];
    if (!v || v.nominalDimensionMm <= 0) {
      throw new Error(`variant set ${vs.id} variant ${i}: nominalDimensionMm must be positive`);
    }
    if (!v.hardwareId.trim()) {
      throw new Error(`variant set ${vs.id} variant ${i}: hardwareId must be non-empty`);
    }
  }
}
